### ISSUE TITLE
Creating extracts for incremental reading

### ISSUE THREAD

**Issue by An-png** [fonol/anki-search-inside-add-card#98]
When doing incremental reading (as I understand it) you have to extract sections of text. Those sections of text are treated as their own "articles", and are therefore added to the reading queue. Flashcards are then created from the extract. This of course doesn't have to be done all at once, in that same way that you don't have to incrementally read the entire PDF at once.

As I understand it, these extracts are meant to have a spaced repetition review schedule for them, so exactly as to how this would be implemented with the current queue system, I don't know. 
 It also seems like you have made quite a few changes to the priority system in one of your branches, so I don't think that comparing what changes should happen (to make extracts viable) to the current queue system isn't a good idea because the current queue system is out of date.

**Comment by fonol** [comment]
That feature branch has already been merged into the main branch, so it is the system which is currently used. I won't implement a spaced scheduling system for the notes as I am quite happy with the priority based one. I also don't really want to copy SMs (text-only) incremental reading system, as my personal reading is like 99% PDFs. I use textual notes too in the queue, but rather to remind me of tasks than to process texts. As for extracts from text notes, the priority queue system IMO works better with single notes where you mark your progress, than with splitting notes up. 
If you want to mark your progess in a text note, maybe you could 
1. insert some kind of marks in the text with the tinyMCE editor  
2. delete already read text (not really a cool solution)

**Comment by p4nix** [comment]
@An-png If you want to emulate SuperMemo's behaviour, you might want to make a deck for incremental reading snippets and copy text excerpts to an appropriate notetype. The IR add-on could supplement this, but as far as I know it does feature a priority assigned queue as well rather than a spaced repetition of the excerpts. I don't know if this can be disabled.

**Comment by An-png** [comment]
@p4nix The incremental reading addon has been abandoned, and is riddled with bugs. In my current collection it doesn't work at all, and creates bugs with image occlusions. Creating a new "collection" removes all these bugs, but I can't seem to fix them for my current collection. 

Regardless, the addon has been abandoned so no bugs will be fixed and no new features will be added.

@fonol 
> I also don't really want to copy SMs (text-only) incremental reading system, as my personal reading is like 99% PDFs.

Extracts could be made from PDFs as well. I feel like it would better represent sections that you want to read and not read. Currently, highlighting is the only way to do this (e.g. red highlights for sections you want to ignore and green for sections that have been distilled into flashcards) and it is more difficult. This is because highlights can be deleted very easily and by accident, and you can't highlight big chunks of text.

Also, you can mark certain sections with the bookmarks (e.g. the "hard", "more cards" bookmark, e.t.c), whereas a page in a PDF is arbritary, it does not represent any single topic or section, making the "read" marker less meaningful.

>  I use textual notes too in the queue, but rather to remind me of tasks than to process texts.

I will  write about textual notes and why I think that one should be able to incrementally read them in the other issue soon.

> As for extracts from text notes, the priority queue system IMO works better with single notes where you mark your progress, than with splitting notes up.

Why exactly is this? I don't see why adding extracts to the queue would be a problem.

Super-memo also uses a priority system (along with spaced reading), so that you can reduce or increase the priority of articles / extracts

**Comment by fonol** [comment]
> Why exactly is this? I don't see why adding extracts to the queue would be a problem.
Because you can freely change the priority of items at any time, and the priority along with the last time seen determines the position in the queue, so you'd have to manually make sure that the priorities of extracts from the same text are not too different, otherwise, an excerpt from a later position in the text could be placed before one with lower priority.

That paper is interesting, I will take a look if I got some time :)

> One Problem I have, and I don't know if this is a bug, is that when I add a PDF that has a higher priority than some of the ones that currently exist, the PDF I just added is always at the bottom of the queue regardless. I have to readjust the priority of every other PDF in order to move the one I just added higher up the queue.

That is no bug, it is the way the queue system works. I wrote a section on it on the add-on description page.

> Also, in the "Queue manager", every article hsa "no priority".

Also no bug, but no good UI design either, admittedly. The slider you see is for the notes on the right side, so if you select one, and then move and release the slider, it is inserted in the queue with the given priority.

**Comment by An-png** [comment]
>      Because you can freely change the priority of items at any time, and the priority along with the last time seen determines the position in the queue, so you'd have to manually make sure that the priorities of extracts from the same text are not too different, otherwise, an excerpt from a later position in the text could be placed before one with lower priority.

That seems fine, to be honest. Incremental reading is not linear reading, so you do not have to read it in order. This is especially true for something like a scientific paper, where you might want to incrementally read the conclusion first. A solution to this could be automatically applying priorities to earlier sections, or even taking the fact that an extract is earlier in a PDF as a third factor in queue placement, along with assigned priority and time since last viewing. Supermemo also uses some sort of priority queue to handle incremental reading and it seems to be working fine.

Furthermore, without extracts, the only way to be sure you have progressed in a PDF is the marking a page as read. However, often a page contains tons of information, so if you complete half a page (which is already a lot), and then move onto the next page, then the addon will think nothing has been done for that page, keeping it high in the queue. 

This also means you are getting further from the incremental reading concept and have to read big chunks of a single article to make significant "progress" with that article and move it down the queue. The is the opposite of incremental reading, where you are supposed to read small chunks of lots of articles in one sitting. 

On top of this, determining what sections you want to work with ("extracts") can only be done with highlights, which is less permanent then actually making extracts.

> That is no bug, it is the way the queue system works. I wrote a section on it on the add-on description page.

However, it seems that the queue is extremely slow moving. It seems to take a few days to move on it's own, and taking into consideration the issue from before where a "page" is actually multiple topics, the queue doesn't rapidly sift through multiple articles in one day, as it should. 

Another question relating to the queue, does it take in to consideration how much you have read. For example, does an article go down the queue faster if two pages have been read instead of one? If not, then this should be the case. I think the queue should be reversed so that instead of articles that have not been read moving up the queue, articles that have been read move down the queue. This way the more you read one article, the faster it goes down the queue.

> Also no bug, but no good UI design either, admittedly. The slider you see is for the notes on the right side, so if you select one, and then move and release the slider, it is inserted in the queue with the given priority.

There seems to be no easy way to see the assigned priority of an article. I would have thought the queue manager would be the best place to show the assigned priority of an article, as it shows all the PDFs in the queue in one place.

Also, I think the priority queue should be accessible in the main screen of the addon somehow. Currently you have to open a PDF in order to access the priority queue.

**Comment by fonol** [comment]
> Furthermore, without extracts, the only way to be sure you have progressed in a PDF is the marking a page as read. However, often a page contains tons of information, so if you complete half a page (which is already a lot), and then move onto the next page, then the addon will think nothing has been done for that page, keeping it high in the queue.


The amount of reading you have done does not influence the placement in the queue. Regardless of whether you read half a page or 10 pages in a session, once you hit 'Done', the scheduling is the same.

> This also means you are getting further from the incremental reading concept and have to read big chunks of a single article to make significant "progress" with that article and move it down the queue. 

Same applies here. You don't have to read any amount to influence the scheduling, the only factors that will be determining how often you see an item are its priority and the time since last seen.

> However, it seems that the queue is extremely slow moving. It seems to take a few days to move on it's own, and taking into consideration the issue from before where a "page" is actually multiple topics, the queue doesn't rapidly sift through multiple articles in one day, as it should.

The pace of the queue is determined by your usage. As the system is based on marking your progress on single big items, rather than completely doing many small items ('extracts'), you basically decide how fast you cycle through the queue by choosing how much time to spend on each item before hitting 'Done' and moving to the next. If you have 20 items in your queue, and spend 30 mins on each before hitting 'Done', you won't see all of them in one day, but I guess that would not change in an extract-based system either. That is just a result of having many items in the queue. Of course, there could be an option to weight the priority factor higher in calculating the position in the queue, and thus allowing an item with a high priority to maybe appear twice or more in a day before a long-time-not-seen item with low priority appears for the first time. Might add that as a config option.

> There seems to be no easy way to see the assigned priority of an article. I would have thought the queue manager would be the best place to show the assigned priority of an article, as it shows all the PDFs in the queue in one place.

You can see it in the bottom bar of the pdf viewer, and in the edit dialog. I might add a display in the Queue manager when I have some time.

> Also, I think the priority queue should be accessible in the main screen of the addon somehow. Currently you have to open a PDF in order to access the priority queue.

Notes -> Queue

**Comment by An-png** [comment]
> before hitting 'Done

oh, you have to hit done to change the placement in the queue! Wow, I was stupid enough to not notice that. No wonder nothing was changing. Just so that people don't fall to the same mistake, can you mention the fact that you need to hit the "Done!" button to reposition it under the IR section of the addon description?

As an addition, could the addon possibly add the amount of time spent on an article automatically after the add window is closed?

> If you have 20 items in your queue, and spend 30 mins on each...

Does this mean that the amount of time spent before hitting done effects the queue? So if I were to read for 30 minutes it would go down the queue more than if I read 5 minutes? 

I think if extracts were to be implemented, this would have to change. What I would envision is that the "last done" element of the queue priority determination is not dependant on the amount of time spent on an article but rather the amount of that article that was converted into extracts. The priority of the extracts would be determined by the amount of an extract that was converted to flashcards (this could be done automatically if someone converts a sentence into a cloze, or manually otherwise). I understand that this would be a lot of work and so I can see why it's not really a priority for you know, but if in the future you had time this might be something worth implementing.

> Notes -> Queue

Yeah, I phrased my request wrong. I meant that there should be a easy way to see the queue manager from the main page. Maybe `Notes -> Queue manager`? Because as of current you have to click on a PDF and then click the *3 horizontal bars* to see the queue manager. This is not very clear.

**Comment by fonol** [comment]
> As an addition, could the addon possibly add the amount of time spent on an article automatically after the add window is closed?

Could you elaborate? You mean like displaying a short notification?

> Does this mean that the amount of time spent before hitting done effects the queue? So if I were to read for 30 minutes it would go down the queue more than if I read 5 minutes?

Nope, I quote myself from above:

> You don't have to read any amount to influence the scheduling, the only factors that will be determining how often you see an item are its priority and the time since last seen.

> I meant that there should be a easy way to see the queue manager from the main page. Maybe Notes -> Queue manager? Because as of current you have to click on a PDF and then click the 3 horizontal bars to see the queue manager. This is not very clear.


True that, I put it on my list.

**Comment by An-png** [comment]
> item are its priority and the time since last seen.

I think that if I read something for 5 pages it should go down the queue more than if I read q page.. right?

Easy way to do this I guess is to check how nuch time has been spent reading an qrticle since the last time "Done!" was presssed.

**Comment by An-png** [comment]
> insert some kind of marks in the text with the tinyMCE editor

Is this highlighting? As I cannot find any other marking system for text notes.

Also, if a marking system exists, can it be used to mark progress in the browser, just as hitting "read" on a page marks progress in the browser (with the 8 bars in the bottom left of a PDF).

**Comment by An-png** [comment]
> I made the extract feature to be able to extract e.g. chapters from a pdf and have them as different items in the queue, so that you can work on them with different priorities and/or schedules.

Currently, regardless of which page you are on, the extractor chooses the page range from `1-1`. I think a useful but small improvement is when you click `[..] -> extract`, the page range starts at the page you are currently on.

I see the use of this if you are, for example, incrementally reading an entire textbook. Treating a entire 200 page textbook as one document would be a ridiculous idea, so in that sense extracting chapters works. Also, in that case, chapters also start at the top of a page, so page splitting is useful.

Comparatively, I wanted a system that  can read Scientific papers incrementally. Scientific Papers are usually PDFs (especially ones you get off Sci-hub), and pages rarely seperate sections of the article. Due to this, I single page can contain multiple topics, making page-by-page extraction not very useful.

As for extracts, I see this as a way of passively reading the article, not having to make flashcards immediately. In **twenty rules for formulating knowledge**, Woz describes how you need to _learn before you understand_. This is a very important concept, and if you do not understand a small section of an article, then you won't make very good flashcards.

I suppose this isn't a problem with clearly and concisely worded textbooks, but by comparison, scientific papers are significantly more difficult to understand, and often require some background knowledge of the subject. Sometimes you have to read on to get important context on what a paragraph means.

For this reason, often I have to spend multiple minutes at a time to create a single note, when instead if there were extracts I could deal with that later and first **understand** the concepts and ideas being presented. It also means I find it difficult to read more than one paragraph of a PDF at a time, and this also takes quite a while.

[From this article](https://masterhowtolearn.wordpress.com/2019/07/11/what-to-do-after-creating-extracts-in-incremental-reading/) : 

> At extract time, you are already forming passive trace memory engrams of the extracted sentence. The optimum strategy then is not to proceed with generating cloze deletions, but to move on to other elements in the queue or to other extracts in the same article (if the high priority of the article justifies it)

I guess you can sort of create "extracts" by highlighting sections (that you want to want to distill to cards) _yellow_, and dealing with them later, however extracts in SM are shown to you on a _spaced schedule_. This way the information stays fresh in your mind, and you can re-read the extract even after you have distilled it into flashcard. This might prompt you to revisit the source PDF, after you have memorised key points of context through flashcards. This aids understanding as well. 

Comparatively, extracts created by "highlighting by yellow" are unlikely to be visited after they have been distilled, and aren't shown on a schedule. You have to decide what what you will do when a PDF is handed to you in this add-on. (Should I highlight more yellow? Should I convert extracts to cards?) 

The Incremental reading add-on by Joseph Lorimer is essentially abandoned, and doesn't work with PDFs. That is why I am so desperate for a alternative incremental reading system here (as this addon is constantly developed (thank you fonol :smile: ) and supports PDFs).

**Comment by p4nix** [comment]
@An-png I am neither a developer, nor have much time on hand at the moment, but I am thinking about the possibility of using a special notetype to automatically open the Add dialog with a linked PDF (extract). Please note that true IR also offers the possibility to edit texts and so on. This is also the reason why PDF is such a bad format, but as scientists, we are tied to them :(

Via manual edition of the ease factor and review date of those items, integration of the card generation process into the card reviewing process could be possible. Maybe this idea would be even better as a standalone add-on though, as it would allow a bit more flexibility (support for webpages, PDF, texts, maybe video).

However, it would probably be better to separate that discussion from this Github issue.

**Comment by fonol** [comment]
@p4nix already mentioned a big technical culprit. 
Regardless of how these extracts would be presented to you (as Anki cards, as items in the queue with a spaced schedule), I think the main problems are in terms of representation, i.e. how is the extract text created/stored best:

Option 1: Copy-paste text from the PDF into an extract note
+ you then have plain text which is easily editable
- Special characters, formulas, and other stuff like images, tables are hard to preserve that way. 
Currently, not even copying plain text works perfectly (no whitespace in the pasted text where there is an end-of-line in the pdf). So with this method, you might spend quite some time going over the copied text and correcting it, and/or inserting missing images and tables with the cut-out tool.

Option 2: Using the cut-out tool to have the extract as an image
+ Quickly done, obviously preserves images/tables/...
- Not editable, no text selection, no highlighting ...

Option 3: Extracting/Cropping the extract part out of the pdf page as a small standalone pdf
+ Would keep everything intact, text selection/highlighting still possible
- not sure if possible at all with any python PDF library

As for presenting the extracts on a spaced schedule, I think @p4ninx idea would work and would have the advantage of not having to code a spaced schedule, on the other hand, it would be more elegant to have everything in the queue: extracts could be handled like queued notes with a schedule, on the day they are due they are put in front of the queue just like scheduled notes, sorted by their priority.

**Comment by p4nix** [comment]
@An-png Just wanted to inform you that I started working on my idea. At the moment, it is not usable though and just a work-in-progress-git-repo. So do not bother with downloading etc. Also, until this thing sees its first release, I am not open for feature requests, so be patient and take a look at it once it is ready.

**Comment by An-png** [comment]
Sorry for my late response.

> Just wanted to inform you that I started working on my idea.

How is your system going to work? This is how I would have envisioned it:

1. In the add dialog, you select a piece of text, and press a button or a shortcut to turn it into a "extract". The section that is "extracted" is highlighted for convenience.
2. This extract is sent to a deck, where it can be reviewed using spaced repetition. With this, extract reviewing and flashcard reviewing is intermingled.
3. You can then highlight a section of text when reviewing an extract. This will on the add dialog again, with the text in a certain field (so that it can be clozed).
4. This way you can add clozes.

How exactly do you envision a extract system working, and how exactly will it work?

Also, is your add-on completely separate from this one? If so, why?


>  I had a nice chat with an avid SM user yesterday, and even he admitted IR extracts don't work that well on textbooks where it is difficult to find chunks of text that you can simply extract and then process later, because they are hard to understand without the surrounding context.

I agree to some extent with this. One workaround is that when you are creating a cloze from an extract, as the add dialog opens, this addon will open the PDF related to the extract

**Comment by p4nix** [comment]
> How is your system going to work? This is how I would have envisioned it:
It's going to work how I envisioned it :P Well, actually:

In the main options, there will be a new toolbar option of my add-on. From there, you will be able to add PDFs (also via bulk-import from Zotero), import wikipedia articles, import YouTube videos or create empty notes for copy-pasting of articles/incremental writing. This will create a regular Anki note with all the meta-data in the fields. You will be able to select the first date for scheduling + the growth rate (ease factor). Content will be saved in the form of HTML (for Wikipedia imports, text files [using tinyMCE, just like this add-on]), as a link (for youtube, maybe iframes of webpages), or as a link to a local file (PDF).

Once the note is reviewed (via the regular Anki scheduler, as of now priorities are not planned to be handled), it will automatically open the Anki Add Cards dialog in a split screen. From there you can create new cards, extract texts to new tinyMCE notes (while generating a knowledge tree like structure via hierarchical tags), modify the content (obviously not possible for PDF), etc. Once you are done with this - if you are lazy just an extract and some clozes in a few seconds - you will save the edited note and schedule it to your liking (by default, exponential growth according to the specified ease factor). The extracts/clozes/etc generated from the notes will also be scheduled for review in the future, or as a learning card (if you are into wasting your time), depending on your preferences.

Based on the hierarchical tags, a knowledge tree similar to SuperMemo will be created, which can be navigated to find other extracts and cards generated from that article and so on (parent, child-relationships)..

> Also, is your add-on completely separate from this one? If so, why?
I need full metadata saved in a database, good support for citation, etc - all of those things would be quite a hassle to add retroactively to the database, but are easily done via Anki note types.
Also, text notes can also be viewed from mobile devices this way. Editing and making cards would be annoying though ;)
The goals just differ too much, and I believe it is faster to setup my own thing without building around the system in a way it is not designed..

> Also, one suggestion (@p4nix) I have is to make the add-on you are working on work with image occlusion, both to create extracts and to create clozes from those extracts.

It is planned to be able to create extracts as marked areas of the PDF, spanning multiple pages (+ sections on single pages), or just section(s) of the PDF (because of science papers using multicol usually). However, as of now, I haven't begun the fight against pdf.js yet, but as I depend on reading science papers just like you (and would appreciate a Supermemo-like process), I can guarantee that this fight will be won :P
As the add-on will build a knowledge tree based on hierarchical tags, any IR cards you generate from the cards itself will be attributed to it. It is thinkable to make a direct integration, which would also allow adding additional meta data to the IO card, but right now I am focusing on getting the basic things to work. But I am keeping it in mind, even if it is just the difference of one shortcut vs two shortcuts + better support for pasting metadata automatically and in a proper way.

I, and probably also fonol, would appreciate if you put discussion about my add-on on my github though. I do not forbid you to dream, but I actually know how IR in SuperMemo works (although I am definitely not experienced) and would rather not be interrupted by some feature requests before base functionality is implemented.

Also, I am a chemistry student, not a coder, and thus will only benefit from features if they save me time regarding my research/knowledge seeking interests, as being able to code doesn't help much in the field of organic chemistry. But as I value myself, the GUI will be kept simple, yet functional, and easy to use for everyone. I am trying my best to code it in a modular way as much as possible. In theory, all data which can be put in an Anki database field and displayed via QT or a browser could be added later: the sky is the limit - but I am currently implementing the ground.

**Comment by An-png** [comment]
> it will automatically open the Anki Add Cards dialog in a split screen. From there you can create new cards, extract texts to new tinyMCE notes

Why? Couldn't the viewer simply show on the face of a card? For extracts, you could have a shortcut to automatically send the highlighted section to a field of a certain note type, like the current incremental reading does.

> good support for citation

What do you mean by this? Is this like Supermemo, when the source of the extract is displayed at the bottom?

>  In the main options, there will be a new toolbar option of my add-on. From there, you will be able to add PDFs (also via bulk-import from Zotero), import wikipedia articles, [...]  The extracts/clozes/etc generated from the notes will also be scheduled for review in the future, or as a learning card (if you are into wasting your time), depending on your preferences.

So as I understand it, compared to the **original incremental reading addon**, this addon has the advantage of being able to work with PDFs & Youtube videos as well as text, and has a knowledge tree. 

>  right now I am focusing on getting the basic things to work. 

:smile: Yeah, I'll keep a list of feature requests, that I might ask for later when you have finished implementing the basics.

**Comment by An-png** [comment]
Hello @p4nix , any updates on the project? It seems from your commit history that you are busy right now. 

You mention that the main article will be stored in a note with a ease factor, however, I'm pretty sure this isn't a good idea. For articles of 1000s of words, you could be doing 10s of reviews for them, so why should the gaps increase? Surely it should be a queue system for the articles like the one with search-inside-add-card add-on? I'm pretty sure only the extracts are meant to have spaced repetition.

I kind of wanted to learn how to program in Anki, where should I start? I don't really want to use proprietary software such as PyCharm, or Microsoft's one (VScode). Is there any alternative? Or do I have to use VScod3?

**Comment by p4nix** [comment]
@An-png 
https://github.com/p4nix/true_incrementalreading/issues/1#issue-664575610

As I would rather not spam fonol's issue section anymore, please continue the discussion over there or find a way for private discussion via E-Mail.
